File: iosApp/Screens/SearchScreen.swift
import SwiftUI
import shared

struct SearchScreen: View {
    let onNavigateToDetails: (Int32) -> Void

    @State private var viewModel = KoinHelperSwift.shared.getSearchViewModel()
    @State private var state = SearchState(
        query: "",
        results: [],
        isSearching: false,
        selectedGenre: nil,
        availableGenres: []
    )
    @State private var showFilters: Bool = false

    var body: some View {
        NavigationView {
            ZStack {
                AppColors.background.edgesIgnoringSafeArea(.all)

                VStack(spacing: 0) {

                    VStack(alignment: .leading, spacing: 10) {
                        CineVerseTextField(
                            placeholder: "Search movies by title...",
                            text: Binding(
                                get: { state.query },
                                set: { viewModel.sendIntent(intent: SearchIntent.UpdateQuery(query: $0)) }
                            ),
                            leadingIcon: "magnifyingglass"
                        )

                        HStack {
                            Button(action: { showFilters.toggle() }, label: {
                                HStack {
                                    Image(systemName: "line.horizontal.3.decrease.circle")
                                    Text("Filter by Genre")
                                }
                                .font(AppTypography.labelSmall)
                                .foregroundColor(state.selectedGenre != nil || showFilters ? AppColors.primary : .gray)
                            })

                            Spacer()

                            if let selected = state.selectedGenre {
                                Text("Filtered by: \(selected.name)")
                                    .font(AppTypography.labelSmall)
                                    .foregroundColor(AppColors.primary)
                                    .onTapGesture {
                                        viewModel.sendIntent(intent: SearchIntent.ApplyFilter(genre: nil))
                                    }
                            }
                        }
                        .padding(.top, 4)

                        if showFilters {
                            ScrollView(.horizontal, showsIndicators: false) {
                                HStack(spacing: 8) {
                                    Button(
                                        action: {
                                            viewModel.sendIntent(intent: SearchIntent.ApplyFilter(genre: nil))
                                        },
                                        label: {
                                            Text("All")
                                                .font(AppTypography.labelSmall)
                                                .padding(.horizontal, 12)
                                                .padding(.vertical, 6)
                                                .background(
                                                    state.selectedGenre == nil
                                                        ? AppColors.primary
                                                        : Color.gray.opacity(0.1)
                                                )
                                                .foregroundColor(state.selectedGenre == nil ? .white : .primary)
                                                .cornerRadius(8)
                                        }
                                    )

                                    ForEach(state.availableGenres, id: \.id) { genre in
                                        Button(
                                            action: {
                                                viewModel.sendIntent(intent: SearchIntent.ApplyFilter(genre: genre))
                                            },
                                            label: {
                                                Text(genre.name)
                                                    .font(AppTypography.labelSmall)
                                                    .padding(.horizontal, 12)
                                                    .padding(.vertical, 6)
                                                    .background(
                                                        state.selectedGenre?.id == genre.id
                                                            ? AppColors.primary
                                                            : Color.gray.opacity(0.1)
                                                    )
                                                    .foregroundColor(
                                                        state.selectedGenre?.id == genre.id ? .white : .primary
                                                    )
                                                    .cornerRadius(8)
                                            }
                                        )
                                    }
                                }
                            }
                            .transition(.slide)
                        }
                    }
                    .padding()
                    .background(AppColors.surface)

                    if state.isSearching {
                        Spacer()
                        ProgressView()
                            .progressViewStyle(CircularProgressViewStyle(tint: AppColors.primary))
                        Spacer()
                    } else if state.results.isEmpty {
                        Spacer()
                        Text(state.query.isEmpty ? "Search for your favorite movies" : "No results found")
                            .foregroundColor(.gray)
                            .font(AppTypography.bodyMedium)
                        Spacer()
                    } else {
                        ScrollView {
                            LazyVStack(spacing: 12) {
                                ForEach(state.results, id: \.id) { movie in
                                    Button(action: { onNavigateToDetails(movie.id) }, label: {
                                        HStack(spacing: 12) {

                                            ZStack {
                                                if let path = movie.posterPath, let url = URL(string: path) {
                                                    AsyncImage(url: url) { phase in
                                                        switch phase {
                                                        case .success(let image):
                                                            image
                                                                .resizable()
                                                                .aspectRatio(contentMode: .fill)
                                                                .frame(width: 60, height: 90)
                                                                .clipped()
                                                        case .failure:
                                                            ImageFallbackView(systemIconName: "film")
                                                                .frame(width: 60, height: 90)
                                                        case .empty:
                                                            ProgressView()
                                                                .progressViewStyle(
                                                                    CircularProgressViewStyle(tint: AppColors.primary)
                                                                )
                                                                .frame(width: 60, height: 90)
                                                        @unknown default:
                                                            ImageFallbackView(systemIconName: "film")
                                                                .frame(width: 60, height: 90)
                                                        }
                                                    }
                                                    .frame(width: 60, height: 90)
                                                    .clipped()
                                                } else {
                                                    ImageFallbackView(systemIconName: "film")
                                                        .frame(width: 60, height: 90)
                                                }
                                            }
                                            .frame(width: 60, height: 90)
                                            .cornerRadius(8)
                                            .clipped()

                                            VStack(alignment: .leading, spacing: 6) {
                                                Text(movie.title)
                                                    .font(AppTypography.bodyMedium)
                                                    .fontWeight(.bold)
                                                    .foregroundColor(.primary)
                                                    .multilineTextAlignment(.leading)
                                                    .lineLimit(2)

                                                Text("Year: \(String(movie.releaseDate.prefix(4)))")
                                                    .font(AppTypography.labelSmall)
                                                    .foregroundColor(.gray)

                                                Text("★ \(String(format: "%.1f", movie.voteAverage))")
                                                    .font(AppTypography.labelSmall)
                                                    .foregroundColor(AppColors.primary)
                                                    .fontWeight(.bold)
                                            }
                                            Spacer()
                                        }
                                        .padding(8)
                                        .background(AppColors.surface)
                                        .cornerRadius(12)
                                        .shadow(color: Color.black.opacity(0.2), radius: 5)
                                    })
                                }
                            }
                            .padding()
                        }
                    }
                }
            }
            .navigationTitle("Smart Search")
        }
        .task {
            for await currentState in viewModel.state {
                if let currentState = currentState { self.state = currentState }
            }
        }
    }
}

File: androidApp/features/search/src/main/kotlin/com/cineverse/android/features/search/SearchScreen.kt
package com.cineverse.android.features.search

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.List
import androidx.compose.material.icons.filled.Search
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.FilterChip
import androidx.compose.material3.FilterChipDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.cineverse.core.designsystem.components.CineVerseTextField
import com.cineverse.presentation.search.SearchIntent
import com.cineverse.presentation.search.SearchViewModel
import com.seiko.imageloader.rememberImagePainter
import org.koin.compose.koinInject

@Composable
fun SearchScreen(
    onNavigateToDetails: (Int) -> Unit,
    modifier: Modifier = Modifier,
    viewModel: SearchViewModel = koinInject()
) {
    val state by viewModel.state.collectAsState()
    var showFilters by remember { mutableStateOf(false) }

    Column(
        modifier = modifier
            .fillMaxSize()
            .background(MaterialTheme.colorScheme.background)
    ) {
        
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .background(MaterialTheme.colorScheme.background)
                .padding(16.dp)
        ) {
            Text(
                text = "Smart Search",
                style = MaterialTheme.typography.headlineLarge,
                modifier = Modifier.padding(bottom = 12.dp)
            )

            CineVerseTextField(
                value = state.query,
                onValueChange = { viewModel.sendIntent(SearchIntent.UpdateQuery(it)) },
                placeholder = "Search movies by title...",
                modifier = Modifier.fillMaxWidth(),
                leadingIcon = {
                    Icon(
                        imageVector = Icons.Default.Search,
                        contentDescription = "Search",
                        tint = MaterialTheme.colorScheme.onSurfaceVariant
                    )
                },
                trailingIcon = {
                    IconButton(onClick = { showFilters = !showFilters }) {
                        Icon(
                            imageVector = Icons.Default.List,
                            contentDescription = "Filters",
                            tint = if (state.selectedGenre != null || showFilters) {
                                MaterialTheme.colorScheme.primary
                            } else {
                                MaterialTheme.colorScheme.onSurfaceVariant
                            }
                        )
                    }
                }
            )

            AnimatedVisibility(visible = showFilters || state.selectedGenre != null) {
                LazyRow(
                    contentPadding = PaddingValues(top = 8.dp),
                    horizontalArrangement = Arrangement.spacedBy(8.dp)
                ) {
                    item {
                        FilterChip(
                            selected = state.selectedGenre == null,
                            onClick = { viewModel.sendIntent(SearchIntent.ApplyFilter(null)) },
                            label = { Text("All") },
                            colors = FilterChipDefaults.filterChipColors(
                                selectedContainerColor = MaterialTheme.colorScheme.primary,
                                selectedLabelColor = Color.White
                            )
                        )
                    }
                    items(state.availableGenres) { genre ->
                        FilterChip(
                            selected = state.selectedGenre == genre,
                            onClick = { viewModel.sendIntent(SearchIntent.ApplyFilter(genre)) },
                            label = { Text(genre.name) },
                            colors = FilterChipDefaults.filterChipColors(
                                selectedContainerColor = MaterialTheme.colorScheme.primary,
                                selectedLabelColor = Color.White
                            )
                        )
                    }
                }
            }
        }

        if (state.isSearching) {
            Box(
                modifier = Modifier.fillMaxSize(),
                contentAlignment = Alignment.Center
            ) {
                CircularProgressIndicator(color = MaterialTheme.colorScheme.primary)
            }
        } else if (state.results.isEmpty()) {
            Box(
                modifier = Modifier.fillMaxSize(),
                contentAlignment = Alignment.Center
            ) {
                Text(
                    text = if (state.query.isEmpty()) "Start typing to explore movies!" else "No movies found matching query",
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
            }
        } else {
            LazyColumn(
                modifier = Modifier.fillMaxSize(),
                contentPadding = PaddingValues(horizontal = 16.dp, vertical = 8.dp),
                verticalArrangement = Arrangement.spacedBy(12.dp)
            ) {
                items(state.results) { movie ->
                    Card(
                        modifier = Modifier
                            .fillMaxWidth()
                            .height(110.dp)
                            .clickable { onNavigateToDetails(movie.id) },
                        shape = RoundedCornerShape(12.dp),
                        colors = CardDefaults.cardColors(
                            containerColor = MaterialTheme.colorScheme.surface
                        ),
                        elevation = CardDefaults.cardElevation(defaultElevation = 2.dp)
                    ) {
                        Row(modifier = Modifier.fillMaxSize()) {
                            
                            Box(
                                modifier = Modifier
                                    .width(74.dp)
                                    .fillMaxHeight()
                                    .clip(RoundedCornerShape(topStart = 12.dp, bottomStart = 12.dp))
                            ) {
                                val painter = rememberImagePainter(movie.posterPath ?: "")
                                Image(
                                    painter = painter,
                                    contentDescription = movie.title,
                                    modifier = Modifier.fillMaxSize(),
                                    contentScale = ContentScale.Crop
                                )
                            }

                            Column(
                                modifier = Modifier
                                    .weight(1f)
                                    .padding(12.dp)
                                    .fillMaxHeight(),
                                verticalArrangement = Arrangement.SpaceBetween
                            ) {
                                Column {
                                    Text(
                                        text = movie.title,
                                        style = MaterialTheme.typography.bodyMedium.copy(fontWeight = FontWeight.Bold),
                                        maxLines = 1,
                                        overflow = TextOverflow.Ellipsis
                                    )
                                    Spacer(modifier = Modifier.height(4.dp))
                                    Text(
                                        text = "Release: ${movie.releaseDate.take(4)}",
                                        style = MaterialTheme.typography.labelSmall,
                                        color = MaterialTheme.colorScheme.onSurfaceVariant
                                    )
                                }
                                Text(
                                    text = "Rating: ★ ${movie.voteAverage}",
                                    style = MaterialTheme.typography.labelSmall.copy(fontWeight = FontWeight.Bold),
                                    color = MaterialTheme.colorScheme.primary
                                )
                            }
                        }
                    }
                }
            }
        }
    }
}
